### PR TITLE
Markdown Jekyll

### LINKED ISSUE #2: nef-doc
Script to generate documentation from `playground page` - compatible with `Jekyll`

### PR DESCRIPTION
**Pull request by miguelangel-dev**
###  References
**Issues:**
* Close #2 

## What is the goal?

Script to generate documentation from `playground page` - compatible with `Jekyll`

## How to test it?

**run**: `nef-markdown-jekyll.swift "~/Desktop/Documentacion/Type Clases.playground/Pages/Intro.xcplaygroundpage" "~/Desktop" "/docs/testing/"`

**note**: use the next file like `playground` to convert to Jekyll's file
[Documentacion.zip](https://github.com/bow-swift/nef/files/2910625/Documentacion.zip)

## Coverage
set up **99.4%**

### CODE REVIEW

File: lib/Markup/Parser/MarkupLexical.swift [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+struct LexicalAnalyzer {
+    private let content: String
+    private var currentIndex: Int
+
+    init(content: String) {
+        self.content = content
+        self.currentIndex = 0
+    }
+
+    mutating func nextToken() -> (token: Token, line: String)? {
```
Comment (truizlop) [inline lib/Markup/Parser/MarkupLexical.swift]: Is it possible to make this side-effect free? It is mutating a state.

File: lib/Markup/Parser/MarkupSyntax.swift [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+    private var tokenizer: LexicalAnalyzer
+    private var openingDelimiters: [Token]
+
+    private init(content: String) {
+        tokenizer = LexicalAnalyzer(content: content)
+        openingDelimiters = []
+    }
+
+    private mutating func parse() -> [Node] {
```
Comment (truizlop) [inline lib/Markup/Parser/MarkupSyntax.swift]: Is it possible to make this side-effect free? It is mutating a state.

File: lib/Markup/Render/MarkupJekyll.swift [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    let permalink: String
+    
+    public init(permalink: String) {
+        self.permalink = permalink
+    }
+
+    public func render(content: String) -> String? {
+        guard let syntaxTree = SyntaxAnalyzer.parse(content: content) else { return nil }
+        let filteredSyntaxTree = syntaxTree.filter { !$0.isHidden }.reduce()
+        //if VERBOSE { syntax.forEach { print($0) } }
```
Comment (truizlop) [inline lib/Markup/Render/MarkupJekyll.swift]: Remove commented line.

File: lib/MarkupTests/MarkupTests.swift [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+
+import XCTest
+@testable import Markup
+
+
+//case nef(command: Nef.Command, [Node])
+//case markup(title: String?, String)
+//case comment(String)
+//case code(String)
+//case unknown(String)
```
Comment (truizlop) [inline lib/MarkupTests/MarkupTests.swift]: Remove commented code.

File: lib/MarkupTests/MarkupTests.swift [hunk 5]
```diff
@@ -0,0 +1,198 @@
+//
+//  MarkupTests.swift
+//  MarkupTests
+//
+//  Created by Miguel Ángel Díaz on 22/02/2019.
+//  Copyright © 2019 47 Degrees. All rights reserved.
+//
```
Comment (truizlop) [inline lib/MarkupTests/MarkupTests.swift]: Remove headers.

File: lib/MarkupTests/MarkupTests.swift [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+//case markup(title: String?, String)
+//case comment(String)
+//case code(String)
+//case unknown(String)
+
+
+
+class MarkupTests: XCTestCase {
+
+    override func setUp() {
```
Comment (truizlop) [inline lib/MarkupTests/MarkupTests.swift]: Remove empty methods.

File: lib/MarkupTests/MarkupTests.swift [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+    func testPlainPlaygroundWithSimpleMarkup_parse_returnsMarkupNode() {
+        let input = "//: This is a test\n"
+        let expected: [Node] = [.markup(description: nil, "This is a test\n")]
+
+        let result = Markup.SyntaxAnalyzer.parse(content: input)
+
+        XCTAssertEqual(result, expected)
+    }
+//
+//    func testLeftDelimiterWithoutRightDelimiter_parse_returnsPlainText() {
```
Comment (truizlop) [inline lib/MarkupTests/MarkupTests.swift]: Remove commented code.

File: markdown/JekyllMarkdown/main.swift [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    func show() {
+        switch self {
+        case .error: printError()
+        case .success: printSuccess()
+        case .help: printHelp()
+        }
+    }
+
+    private func printError() {
+        print("ERROR")
```
Comment (truizlop) [inline markdown/JekyllMarkdown/main.swift]: Shouldn't we provide more info about this? Same in `printSuccess` and `printHelp`.

File: lib/Markup/Render/MarkupJekyll.swift [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+        case .hidden:
+            return ""
+        case .invalid:
+            fatalError("Found .invalid command in nef: \(nodes).")
+        }
+    }
+}
+
+extension Node.Code {
+    func jekyll() -> String {
```
Comment (truizlop) [inline lib/Markup/Render/MarkupJekyll.swift:74]: `jekyll` seems like something that should be in a protocol, and the extension of `Node.Code` and `Node.Nef.Command` should be conforming to that protocol. I see the signatures are different in the two cases though. Can we think of a way of unifying them?

**Review by truizlop — CHANGES_REQUESTED** [review]

File: markdown/JekyllMarkdown/main.swift (hunk 8 above)
Comment (miguelangel-dev) [reply markdown/JekyllMarkdown/main.swift]: Yep, sorry, I have not finished this.

File: lib/MarkupTests/MarkupTests.swift (hunk 7 above)
Comment (miguelangel-dev) [reply lib/MarkupTests/MarkupTests.swift]: Yep, sorry, I have not finished this - I have to complete the rest of the unit-tests

File: lib/MarkupTests/MarkupTests.swift (hunk 6 above)
Comment (miguelangel-dev) [reply lib/MarkupTests/MarkupTests.swift]: sure

File: lib/MarkupTests/MarkupTests.swift (hunk 5 above)
Comment (miguelangel-dev) [reply lib/MarkupTests/MarkupTests.swift]: ups! Thanks :)

File: lib/MarkupTests/MarkupTests.swift (hunk 4 above)
Comment (miguelangel-dev) [reply lib/MarkupTests/MarkupTests.swift]: yep

File: lib/Markup/Render/MarkupJekyll.swift (hunk 3 above)
Comment (miguelangel-dev) [reply lib/Markup/Render/MarkupJekyll.swift]: I was thinking to allow it as argument from the command line...but yes, we have to remove it

File: lib/Markup/Parser/MarkupLexical.swift (hunk 1 above)
Comment (miguelangel-dev) [reply lib/Markup/Parser/MarkupLexical.swift]: In the end, we are using it as a scanner for content (like MarkupSyntax) - if we want to do that, we should change approach, to scan the next token given a content + position, or return a new struct with the new state.

Both options, I feel add extra complex to code, so I decided to keep this version saving only the current position for lexical analyzer (making mutating syntax analyzer too - it is using it).

Now, I finished the analyzer, I gonna go to review it again and see if it is possible without mess up the code :)

File: lib/Markup/Render/MarkupJekyll.swift (hunk 9 above)
Comment (miguelangel-dev) [reply lib/Markup/Render/MarkupJekyll.swift:74]: Good point! :) I will review it

**Review by truizlop — APPROVED** [review]

**Pull request merged**